src/TestCase.php

```diff
@@ -2,145 +2,110 @@
 use Asynit\Assert\AssertWebCaseTrait;
-use Asynit\HttpClient\ArtaxAsyncAdapter;
-use Http\Client\HttpAsyncClient;
-use Http\Message\MessageFactory;
+use Psr\Http\Client\ClientInterface;
+use Psr\Http\Message\RequestFactoryInterface;
+use Amp\Sync\Semaphore;
 use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamFactoryInterface;
 
 class TestCase
 {
     use AssertWebCaseTrait;
 
-    /** @var MessageFactory */
-    private $messageFactory;
-
-    /** @var Semaphore */
-    private $semaphore;
-
-    /** @var HttpAsyncClient */
-    private $client;
-
-    final public function __construct(MessageFactory $messageFactory, Semaphore $semaphore, Test $test)
+    final public function __construct(
+        private RequestFactoryInterface $requestFactory,
+        private StreamFactoryInterface $streamFactory,
+        private Semaphore $semaphore,
+        Test $test,
+        protected ClientInterface $client,
+    )
     {
-        $this->messageFactory = $messageFactory;
-        $this->semaphore = $semaphore;
         $this->test = $test;
     }
 
+    public function initialize()
+    {
+        $this->client = $this->setUp($this->client);
+    }
+
     /**
      * Run before each test.
      *
      * Allow to set default services and context, and also decorate the http async client.
      *
-     * @return \Generator|Promise|HttpAsyncClient
+     * @return ClientInterface
      */
-    public function setUp(HttpAsyncClient $asyncClient)
+    public function setUp(ClientInterface $asyncClient): ClientInterface
     {
         return $asyncClient;
     }
 
-    final public function initialize(bool $allowSelfSignedCertificate = false)
-    {
-        return call(function () use ($allowSelfSignedCertificate) {
-            $this->client = yield call(function () use ($allowSelfSignedCertificate) {
-                if ($allowSelfSignedCertificate) {
-                    $tlsContext = new ClientTlsContext();
-                    $tlsContext = $tlsContext->withoutPeerVerification();
-                }
-
-                return $this->setUp(new ArtaxAsyncAdapter($this->messageFactory, new DefaultClient(null, null, $tlsContext ?? null)));
-            });
-        });
-    }
-
     /**
      * Allow to test a rejection or a resolution of an async call.
      */
-    final protected function sendRequest(RequestInterface $request): Promise
+    final protected function sendRequest(RequestInterface $request): ResponseInterface
     {
-        return \Amp\call(function () use ($request) {
-            /** @var Lock $lock */
-            $lock = yield $this->semaphore->acquire();
-            $response = yield $this->client->sendAsyncRequest($request);
+        $lock = $this->semaphore->acquire();
+        $response = $this->client->sendRequest($request);
+        $lock->release();
 
-            $lock->release();
-
-            return $response;
-        });
+        return $response;
```

Comment:
Actually, I would provide a "client" that wrap this logic to the test case.
1. it's cleaner (no need for this hack `FINAL protected`, same for all other methods)
2. No need to inject the semaphore
3. No need to inject the request factory
4. No need to inject the steamFactory
5. No need to inject the "raw" HttpClient client